<a href="https://colab.research.google.com/github/Kanyawesr/data_viz_2022/blob/main/Pandas101.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [134]:
import pandas as pd

.csv = Comma Separated Values ไฟล์ที่ใช้ (,) คอมม่าในการแยกค่าแต่ละค่า

data.go.th (แหล่งข้อมูลส่วนใหญ่)

In [135]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [136]:
import os # os import เพื่อจัดการเกี่ยวกับพวกไฟล์ต่างๆ ว่าอยู่ในโฟลเดอร์ไหน

In [137]:
path = '/content/drive/My Drive/dataviz_2021_data' #ชี้แหล่งที่มาหลังจากเชื่อมเสร็จ

In [138]:
covid_file_path = os.path.join(path,'pm-20-jan-2021.csv') #os.path.join จะเอา part ที่อยู่ในโฟลเดอร์มาเชื่อมกับ pm-20-jan-2021.csv 
print(covid_file_path) #ความสำคัญของมันคือ part มันจะแตกต่างกัน เช่น window (run คนละ os มันก็จะไม่เหมือนกัน)

/content/drive/My Drive/dataviz_2021_data/pm-20-jan-2021.csv


In [139]:
a = path + '/pm-18-jan-2021.csv' 
print(a)

/content/drive/My Drive/dataviz_2021_data/pm-18-jan-2021.csv


## load data to memory

In [140]:
data_covid = pd.read_csv(covid_file_path) #แสดงผลใช้ csv โดยเป็นคำสั่ง pd.read_csv 
data_covid #เอามาใส่ตัวแปร data_covid และแสดงผล 

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 10 columns]

In [141]:
data_covid.head() #เพื่อดูหัวตารางโดยโหลดข้อมูลให้น้อยที่สุด สมารถกำหนด rows ได้ในวงเล็บถ้าไม่ใว่ก็จะมีแค่ 5 rows
#.head ใช้ได้กับ DataFrame(ตัวแปร)

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[5 rows x 10 columns]

NaN = Not a Number ข้อมูลที่ไม่มีค่า

In [142]:
data_covid.head(20) #แสดง 20 rows

No.  ...                                               risk
0     1  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
1     2  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
2     3  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
3     4  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
4     5  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
5     6  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
6     7  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
7     8  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
8     9  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
9    10  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
10   11  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11   12  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12   13  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
13   14  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
14   15  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
15   16  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...
16   17  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
17   18  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
18   19  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
19   20  ...                      คนไทยเดินทางกลับจากต่างประเทศ

[20 rows x 10 columns]

In [143]:
data_covid.tail() #แสดงข้อมูลท้ายตาราง

No.  ...                                         risk
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[5 rows x 10 columns]

In [144]:
data_covid.shape #เพื่อแสดงผลว่าตารางมีกี rows, columns

(12653, 10)

## ชี้ค่าในตาราง

#### basic

In [145]:
data_covid['province_of_onset'] #แสดงเฉพาะหัวข้อนั้นๆ

0        กรุงเทพมหานคร
1        กรุงเทพมหานคร
2               นครปฐม
3        กรุงเทพมหานคร
4        กรุงเทพมหานคร
             ...      
12648           ชลบุรี
12649            ระยอง
12650            ระยอง
12651            ระยอง
12652              ตาก
Name: province_of_onset, Length: 12653, dtype: object

In [146]:
data_covid['sex']

0        หญิง
1        หญิง
2        หญิง
3         ชาย
4        หญิง
         ... 
12648    หญิง
12649    หญิง
12650    หญิง
12651    หญิง
12652    หญิง
Name: sex, Length: 12653, dtype: object

In [147]:
data_covid['risk'][12649] #ชี้ด้วยชื่อคอลัมและตามด้วยตำแหน่งของ rows

'สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้'

In [148]:
data_covid['sex'][12648]

'หญิง'

#### .iloc

In [149]:
data_covid.iloc[4,7] #เปลี่ยนมุมมองของตารางให้เป็น array ได้โดยใช้ .ilo ตามด้วย [แถว,หลัก]

'กรุงเทพมหานคร'

In [150]:
data_covid.iloc[7,7]

'ประจวบคีรีขันธ์'

# เรียนถึงนี่ 12 Jan 2022

## Table slicing

#### ตัดมาเฉพาะ column ที่ต้องการ

In [151]:
smaller_table = data_covid[['province_of_onset','announce_date','risk']]
smaller_table #เลือกหัวข้อตารางที่ต้องการ

province_of_onset  ...                                         risk
0         กรุงเทพมหานคร  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1         กรุงเทพมหานคร  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2                นครปฐม  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3         กรุงเทพมหานคร  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4         กรุงเทพมหานคร  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...                 ...  ...                                          ...
12648            ชลบุรี  ...                             State Quarantine
12649             ระยอง  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650             ระยอง  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651             ระยอง  ...                                Cluster ระยอง
12652               ตาก  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 3 columns]

In [152]:
smaller_table1 = data_covid[['sex','age','nationality']]
smaller_table1 

sex   age nationality
0      หญิง  61.0       China
1      หญิง  74.0       China
2      หญิง  73.0    Thailand
3       ชาย  68.0       China
4      หญิง  66.0       China
...     ...   ...         ...
12648  หญิง  44.0    Thailand
12649  หญิง  52.0    Thailand
12650  หญิง  23.0    Thailand
12651  หญิง  29.0    Thailand
12652  หญิง  22.0       Burma

[12653 rows x 3 columns]

#### ตัดเฉพาะ rows ที่ต้องการ

##### แบบง่าย แต่ใช้การอะไรไม่ค่อยจะได้ (มองเป็น array แล้วตัดแบบ array)

$pd_df$[list_of_logical_values] คือการเลือก row ที่ logical values เป็น True


In [153]:
data_covid['province_of_onset'] #เลือกคอลัม province of onset

0        กรุงเทพมหานคร
1        กรุงเทพมหานคร
2               นครปฐม
3        กรุงเทพมหานคร
4        กรุงเทพมหานคร
             ...      
12648           ชลบุรี
12649            ระยอง
12650            ระยอง
12651            ระยอง
12652              ตาก
Name: province_of_onset, Length: 12653, dtype: object

In [154]:
data_covid['province_of_onset']=='ขอนแก่น' #ทำ logical values โดยเครื่องหมาย == คือ if ค่ามันเท่ากับขอนแก่น
#แสดงผลทั้งหมดไม่รู้ true อยู่ row ไหน

0        False
1        False
2        False
3        False
4        False
         ...  
12648    False
12649    False
12650    False
12651    False
12652    False
Name: province_of_onset, Length: 12653, dtype: bool

In [155]:
data_covid[data_covid['province_of_onset']=='ขอนแก่น'] # โดยมาจาก list_of_logical_values ก็จะแสดงแต่ค่า true

No.  ...                                           risk
180      181  ...                                        สนามมวย
462      463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466    1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970    1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637    2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ
2673    2674  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
5948    5949  ...                              Cluster สมุทรสาคร
6082    6083  ...                              Cluster สมุทรสาคร
9333    9334  ...                           อยู่ระหว่างการสอบสวน
9334    9335  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
10610  10611  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11517  11518  ...                           อยู่ระหว่างการสอบสวน
11697  11698  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11698  11699  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[14 rows x 10 columns]

In [156]:
data_covid[data_covid['age']>=70] #เลือกอายุที่มากกว่าหรือเท่ากับ 70

No.  ...                                               risk
1          2  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
7          8  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
24        25  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...
82        83  ...                                            สนามมวย
...      ...  ...                                                ...
11694  11695  ...      ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
11700  11701  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11730  11731  ...                                  Cluster สมุทรสาคร
11741  11742  ...                               อยู่ระหว่างการสอบสวน
12102  12103  ...                                  Cluster สมุทรสาคร

[245 rows x 10 columns]

In [157]:
data_covid[data_covid['nationality']=='Thailand'] #เลือกเฉพาะคนไทย

No.  ...                                               risk
2          3  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
15        16  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...
19        20  ...                      คนไทยเดินทางกลับจากต่างประเทศ
20        21  ...                      คนไทยเดินทางกลับจากต่างประเทศ
22        23  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...
...      ...  ...                                                ...
12647  12648  ...      ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
12648  12649  ...                                   State Quarantine
12649  12650  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                      Cluster ระยอง

[7313 rows x 10 columns]

In [158]:
data_covid.iloc[1:5,:] #การเลือกเฉพาะ rows ที่ต้องการ

No. announce_date  ... district_of_onset                              risk
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[4 rows x 10 columns]

แบบ advanced แต่ใช้ประโยชน์ได้เยอะกว่า (ตัดแบบใช้ logic query)

$pddf$[list_of_logical_values] คือการเลือก row ที่ logical values เป็น True

In [159]:
data_covid['province_of_onset']

0        กรุงเทพมหานคร
1        กรุงเทพมหานคร
2               นครปฐม
3        กรุงเทพมหานคร
4        กรุงเทพมหานคร
             ...      
12648           ชลบุรี
12649            ระยอง
12650            ระยอง
12651            ระยอง
12652              ตาก
Name: province_of_onset, Length: 12653, dtype: object

In [160]:
data_covid['province_of_onset']=='ขอนแก่น'

0        False
1        False
2        False
3        False
4        False
         ...  
12648    False
12649    False
12650    False
12651    False
12652    False
Name: province_of_onset, Length: 12653, dtype: bool

In [161]:
data_covid[data_covid['province_of_onset']=='ขอนแก่น']

No.  ...                                           risk
180      181  ...                                        สนามมวย
462      463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466    1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970    1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637    2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ
2673    2674  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
5948    5949  ...                              Cluster สมุทรสาคร
6082    6083  ...                              Cluster สมุทรสาคร
9333    9334  ...                           อยู่ระหว่างการสอบสวน
9334    9335  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
10610  10611  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11517  11518  ...                           อยู่ระหว่างการสอบสวน
11697  11698  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11698  11699  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[14 rows x 10 columns]

In [162]:
data_covid[data_covid['age']>=70]

No.  ...                                               risk
1          2  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
7          8  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
24        25  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...
82        83  ...                                            สนามมวย
...      ...  ...                                                ...
11694  11695  ...      ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
11700  11701  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11730  11731  ...                                  Cluster สมุทรสาคร
11741  11742  ...                               อยู่ระหว่างการสอบสวน
12102  12103  ...                                  Cluster สมุทรสาคร

[245 rows x 10 columns]

การชี้ค่าในตารางด้วย data_covid[***x***] ค่า ***x*** สามารถมีค่าที่ใส่เข้าไปได้ 2 แบบ


1.   ชื่อ column
2.   ใส่ rows ที่ต้องการ (ใส่ list ที่มี***ขนาดเท่ากับจำนวนแถว*** แล้วบอกว่าเอา แถวไหนบ้างโดยใช้ logical expression (```True```/```False```))



In [163]:
eight_rows_covid = data_covid.iloc[:8,3:] # ตารางตัวอย่าง โดยเลือกมาแค่ 8 row 
eight_rows_covid

sex   age  ... district_of_onset                              risk
0  หญิง  61.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1  หญิง  74.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2  หญิง  73.0  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3   ชาย  68.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4  หญิง  66.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5  หญิง  33.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
6  หญิง  57.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
7  หญิง  73.0  ...            หัวหิน  คนต่างชาติเดินทางมาจากต่างประเทศ

[8 rows x 7 columns]

In [164]:
eight_rows_covid[[True,False,True,True,False,True,False,False]]   #2.   ใส่ list ที่ มีค่า logic ของ แต่ละ row โดยที่ ค่า True คือ เลือก ค่า False คือไม่เลืิอก

sex   age  ... district_of_onset                              risk
0  หญิง  61.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2  หญิง  73.0  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3   ชาย  68.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5  หญิง  33.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[4 rows x 7 columns]

[True,False,True,True,False,True,False,False] เลือก list จากตารางที่กำหนดมา(eight_rows_covid) ว่าจะเอาตัวไหนบ้าง(True)

In [165]:
eight_rows_covid['province_of_onset'] == 'กรุงเทพมหานคร'  #สร้าง list ของ logical expression

0     True
1     True
2    False
3     True
4     True
5     True
6     True
7    False
Name: province_of_onset, dtype: bool

In [166]:
eight_rows_covid[eight_rows_covid['province_of_onset'] == 'กรุงเทพมหานคร']  #เราเอา 2 บรรทัดบนมาบวกกัน

sex   age  ... district_of_onset                              risk
0  หญิง  61.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1  หญิง  74.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
3   ชาย  68.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4  หญิง  66.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5  หญิง  33.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
6  หญิง  57.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[6 rows x 7 columns]

In [167]:
eight_rows_covid['district_of_onset'] == 'เมือง' #หากคอลัมมีค่า NaN มันก็จะเป็น False 

0    False
1    False
2     True
3    False
4    False
5    False
6    False
7    False
Name: district_of_onset, dtype: bool

In [168]:
eight_rows_covid['district_of_onset'] == 'หัวหิน'

0    False
1    False
2    False
3    False
4    False
5    False
6    False
7     True
Name: district_of_onset, dtype: bool

In [169]:
import numpy as np

$.isnull()$ ใช้ตรวจสอบว่าเป็น NaN หรือไม่

เพื่อตามหาค่าว่าง(NaN) ให้แสดงผลเป็น True

In [170]:
eight_rows_covid

sex   age  ... district_of_onset                              risk
0  หญิง  61.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1  หญิง  74.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2  หญิง  73.0  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3   ชาย  68.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4  หญิง  66.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5  หญิง  33.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
6  หญิง  57.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
7  หญิง  73.0  ...            หัวหิน  คนต่างชาติเดินทางมาจากต่างประเทศ

[8 rows x 7 columns]

In [171]:
eight_rows_covid['district_of_onset'].isnull()

0     True
1     True
2    False
3     True
4     True
5     True
6     True
7    False
Name: district_of_onset, dtype: bool

Quiz (ให้ใช้เฉพาะที่อาจารย์สอนไปแล้วในวิชานี้)

1.   คำนวณ อายุเฉลี่ย ของ**ผู้ชาย** ในข้อมูลทั้งหมด
2.   คำนวณ อายุเฉลี่ย ของ**ผู้ชาย** ในจังหวัดขอนแก่น
3.   หาจำนวนผู้ป่วยที่เป็นคน **ที่ไม่ใช่คนไทย**

In [172]:
this_data = data_covid[['sex','age','province_of_onset']] #กำหนดคอลัมเลือก เพศ อายุ จังหวัด

In [173]:
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [174]:
female = this_data[this_data['sex']=='หญิง'] #สร้างตัวแปร เลือกเฉพาะที่เป็นเพศหญฺง
female['age'] #พร้อมกับแสดงคอลัมอายุ

0        61.0
1        74.0
2        73.0
4        66.0
5        33.0
         ... 
12648    44.0
12649    52.0
12650    23.0
12651    29.0
12652    22.0
Name: age, Length: 5332, dtype: float64

In [175]:
sum = 0 #ให้ sum เป็นตัวแปรโดยกำหนดให้เท่ากับ 0 เพราะถ้าเป็นค่าอื่นเลขมันจะเพิ่มขึ้นตามเลขที่เรากำหนด
N = 0 #เป็นตัวแปรเท่ากับ 0
for a in female['age']:
    
    if a > 0: #ถ้าอายุ(a)มากกว่า 0
        sum += a # sum = sum + a (ผลรวมอายุทั้งหมดมาบวกกับ a เรื่อยๆ) 
        N += 1 #คำนวณจำนวนโดยบวก 1 ไปเรื่อยๆ มี 1 ตัว 2,3,4,5 ตัวบวก 1 ไปเรื่อยๆ

print(f'อายุเฉลี่ย ของ ผู้ป่วยหญิง {sum/N}')

อายุเฉลี่ย ของ ผู้ป่วยหญิง 36.63955450022614


In [176]:
female_KK = female[female['province_of_onset']=='ขอนแก่น']

sum = 0
N = 0
for a in female_KK['age']:
    
    if a > 0:
        sum += a # sum = sum + a
        N += 1

print(f'อายุเฉลี่ย ของ ผู้ป่วยหญิง ในจังหวัดขอนแก่น {sum/N}')

อายุเฉลี่ย ของ ผู้ป่วยหญิง ในจังหวัดขอนแก่น 41.57142857142857


1.คำนวณ อายุเฉลี่ย ของผู้ชาย ในข้อมูลทั้งหมด

In [177]:
this_data = data_covid[['sex','age','province_of_onset']] #เลือกคอลัมที่ต้องการคือ เพศ อายุ และจังหวัด
this_data #แสดงผล

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [178]:
male = this_data[this_data['sex']=='ชาย'] #เลือกมาเฉพาะแต่เพศชายโดยเลือกมาจากตาราง this_data
male['age']

3        68.0
8        63.0
10       33.0
11       61.0
12        6.0
         ... 
12640    66.0
12641    29.0
12643    39.0
12646    18.0
12647    42.0
Name: age, Length: 4819, dtype: float64

In [179]:
sum = 0 #มีตัวแปร sum มารับให้เริ่มเท่ากับ 0 ถ้าเริ่มค่าอื่นไม่ได้ /ถ้าเป็นคูณเริ่มที่ 1
N = 0 #เอาตัวเลขทั้งหมด 
for a in male['age']: 
    
    if a > 0: #เลือกมากกว่า 0 เพื่อหลีกเลี่ยงค่าผิดปกติ
        sum += a # sum = sum + a  #เอา a มาบวก sum เรื่อยๆ
        N += 1 

print(f'อายุเฉลี่ย ของ ผู้ป่วยชาย {sum/N}') #อายุเฉลี่ย หาร จำนวน

อายุเฉลี่ย ของ ผู้ป่วยชาย 39.46076940331987


2.คำนวณ อายุเฉลี่ย ของผู้ชายในจังหวัดขอนแก่น

In [180]:
male_KK = male[male['province_of_onset']=='ขอนแก่น'] #สนใจข้อมูลเฉาะที่เป็นจังหวัดขอนแก่น
male_KK

sex   age province_of_onset
180    ชาย  33.0           ขอนแก่น
1466   ชาย  19.0           ขอนแก่น
2673   ชาย  68.0           ขอนแก่น
9333   ชาย  17.0           ขอนแก่น
9334   ชาย  20.0           ขอนแก่น
10610  ชาย  17.0           ขอนแก่น
11517  ชาย  12.0           ขอนแก่น

In [181]:
sum = 0
N = 0
for a in male_KK['age']: #วนดูค่าอายุที่อยู่ในคอลัม
    
    if a > 0:
        sum += a # sum = sum + a
        N += 1

print(f'อายุเฉลี่ย ของ ผู้ป่วยชาน ในจังหวัดขอนแก่น {sum/N}')

อายุเฉลี่ย ของ ผู้ป่วยชาน ในจังหวัดขอนแก่น 26.571428571428573


3.หาจำนวนผู้ป่วยที่ไม่ใช่คนไทย

In [182]:
data_covid[data_covid['nationality'] != 'Thailand']

No.  ...                                          risk
0          1  ...              คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...              คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...              คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...              คนต่างชาติเดินทางมาจากต่างประเทศ
5          6  ...              คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                           ...
12614  12615  ...                       Cluster โรงงาน Big Star
12615  12616  ...                       Cluster โรงงาน Big Star
12617  12618  ...  ผู้ที่เดินทางมาจากต่างประเทศ และเข้า ASQ/ALQ
12618  12619  ...  ผู้ที่เดินทางมาจากต่างประเทศ และเข้า ASQ/ALQ
12652  12653  ...   สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[5340 rows x 10 columns]

โจทย์กำหนดเอง คำนวณอายุเฉลี่ยของผู้ป่วยต่างชาติที่มีอายุน้อยกว่าหรือเท่ากับ 30 ปี

In [183]:
aewymoly = data_covid[data_covid['nationality'] != 'Thailand']
aewymoly

No.  ...                                          risk
0          1  ...              คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...              คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...              คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...              คนต่างชาติเดินทางมาจากต่างประเทศ
5          6  ...              คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                           ...
12614  12615  ...                       Cluster โรงงาน Big Star
12615  12616  ...                       Cluster โรงงาน Big Star
12617  12618  ...  ผู้ที่เดินทางมาจากต่างประเทศ และเข้า ASQ/ALQ
12618  12619  ...  ผู้ที่เดินทางมาจากต่างประเทศ และเข้า ASQ/ALQ
12652  12653  ...   สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[5340 rows x 10 columns]

In [184]:
aewymoly = aewymoly[['sex','age','nationality']]
aewymoly

sex   age nationality
0      หญิง  61.0       China
1      หญิง  74.0       China
3       ชาย  68.0       China
4      หญิง  66.0       China
5      หญิง  33.0       China
...     ...   ...         ...
12614  หญิง  43.0       Burma
12615  หญิง  35.0       Burma
12617  หญิง  12.0     Ireland
12618  หญิง  34.0    Pakistan
12652  หญิง  22.0       Burma

[5340 rows x 3 columns]

In [185]:
sum = 0
N = 0
for ae in aewymoly['age']:
    
    if ae <= 30:
        sum += ae 
        N += 1

print(f'อายุเฉลี่ย ของผู้ป่วยต่างชาติที่มีอายุน้อยกว่าหรือเท่ากับ 30 ปี {sum/N}')

อายุเฉลี่ย ของผู้ป่วยต่างชาติที่มีอายุน้อยกว่าหรือเท่ากับ 30 ปี 23.86591013824885


การจัดการ Missing Value

- ลบ record ที่เป็น missing 
- แทนที่ ค่า missing ด้วยค่าที่เหมาะสม mean, default, category-unknown
- ใช้ ค่าจาก columns อื่นๆ ช่วยประมาณค่า ค่าใน column ที่หายไป (regression, deep learning, etc.)

### ลบ record (dropna)

missing = ```None``` ```NA```(not autorized) ```NaN``` (not a number)

In [212]:
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [186]:
this_data.shape #(12423, 3)

(12653, 3)

In [213]:
this_data.dropna() #จาก 12423 เหลือ 8478 rows ตัดค่า missing

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[8478 rows x 3 columns]

In [214]:
print(data_covid.shape)
print(data_covid.dropna().shape) #ยิ่งคอลัมเยอะ ยิ่งมีค่า missing มากขึ้น

(12653, 10)
(6655, 10)



*   พิจาราณา 3 คอลัม Data เหลือประมาณ 8478
*   พิจารณา 10 คอลัม Data เหลือประมาณ 6655

(Data หายไปประมาณครึ่งนึง)



In [215]:
this_data.dropna(subset=['age','sex']).shape #dropna มีพารามิเตอร์เพิ่มมา 1 ก็คือ(subset=[]) โดยจะให้ตัวโปรแกรมพิจารณาว่าจะเลือกคอลัมไหน

(8880, 3)

In [216]:
print(this_data.dropna().shape)
print(this_data.dropna(subset=['age','sex']).shape) 
#เมื่อกำหนดให้พิจารณาที่ค่าต้องการผลปรากฏว่าค่า missing หายน้อยกว่าเดิม
#dropna เฉพาะคอลัมประมวลผลที่เราสนใจ

(8478, 3)
(8880, 3)


# ตรวจสอบ Missing (```isnull()```)

In [217]:
this_data.isnull() #ตรวจสอบค่า missing ถ้าเป็นจะขึ่น True
#แต่ไม่สามารถตรวจสอบอย่างละเอียดได้

sex    age  province_of_onset
0      False  False              False
1      False  False              False
2      False  False              False
3      False  False              False
4      False  False              False
...      ...    ...                ...
12648  False  False              False
12649  False  False              False
12650  False  False              False
12651  False  False              False
12652  False  False              False

[12653 rows x 3 columns]

In [218]:
this_data.isnull().any() # .any() เอาค่าความจริงในคอลัมทั้งหมดมา OR กัน
#ถ้ามีค่า true สักค่านึงในนี้แล้วไป or กับ false ก็จะเป็น true ก็คือค่า missing

sex                  True
age                  True
province_of_onset    True
dtype: bool

In [219]:
this_data.isnull().all() #เอาค่าความจริงมา AND กัน (ไม่ค่อยได้ใช้งาน)

sex                  False
age                  False
province_of_onset    False
dtype: bool

In [220]:
this_data.isnull().any().any() # เอาทุกอย่างมา OR กันหมดเลยทำห้ทราบว่าตารางนี้มีค่าว่าง

True

In [221]:
this_data.dropna(subset=['sex','age']).isnull().any() # sex,age แก้ค่า missing แล้วเลยเป็นค่า false

sex                  False
age                  False
province_of_onset     True
dtype: bool

In [223]:
this_data.dropna(subset=['sex','age'])

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[8880 rows x 3 columns]

In [222]:
this_data.isnull().any() #ปัญหาคือยังมีค่า missing อยู่

sex                  True
age                  True
province_of_onset    True
dtype: bool

###1.   เพื่อให้ได้ผลลัพธ์การจัดการ ```dropna()``` ต้องเอาตัวแปรมารับ




In [224]:
this_data_nonull = this_data.dropna(subset=['sex','age'])

In [225]:
this_data_nonull.isnull().any() #จัดการค่า missing สำเร็จ

sex                  False
age                  False
province_of_onset     True
dtype: bool

###2.   update ตารางด้วยตัวแปร ```inplace```

(ไม่แนะนำ แต่เพื่อที่จะได้ไม่งงเวลากลับไปอ่านโค้ดอื่นตัวนี้ใส่มาเพื่ออะไร)




In [226]:
this_data.dropna(subset=['sex','age'],inplace=True) #ไปอัปเดตตาราง เมื่อ run ออกมาจไม่แสดงค่า แต่จะขึ้นข้อความเตือน

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [227]:
this_data.isnull().any() #ตารางอัปเดต

sex                  False
age                  False
province_of_onset     True
dtype: bool

In [228]:
aa = this_data.dropna(subset=['sex','age'],inplace=True) #ไม่ควรเอาตัวแปรไปรับ inplace

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [229]:
aa

### แทน missing ด้วยค่าที่เหมาะสม (```fillna```)

In [233]:
this_data = data_covid[['sex','age','province_of_onset']]

In [234]:
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [235]:
print(f'จำนวน record ก่อน drop missing ใน province {this_data.shape[0]}')
print(f'จำนวน record หลัง drop missing ใน province {this_data.dropna(subset=["province_of_onset"]).shape[0]}')

จำนวน record ก่อน drop missing ใน province 12653
จำนวน record หลัง drop missing ใน province 10662


###1. แทนด้วย class ใหม่

In [236]:
this_data_updated = this_data.fillna(value={'sex':'ไม่รู้', 'age':-99, 'province_of_onset':'ไม่รู้'}) #เป็นค่าที่ไม่ได้เป็นตัวเลขอาจแทนด้วย 'ไม่รู้' แต่ก็จะกลายเป็นค่าใหม่ 
this_data_updated.head() 

sex   age province_of_onset
0  หญิง  61.0     กรุงเทพมหานคร
1  หญิง  74.0     กรุงเทพมหานคร
2  หญิง  73.0            นครปฐม
3   ชาย  68.0     กรุงเทพมหานคร
4  หญิง  66.0     กรุงเทพมหานคร

In [238]:
this_data_updated.tail(15)  #ค่าที่กำหนดขึ้นมาได้แทนลงแล้ว

sex   age province_of_onset
12638  ไม่รู้ -99.0         สมุทรสาคร
12639  ไม่รู้ -99.0         สมุทรสาคร
12640     ชาย  66.0   พระนครศรีอยุธยา
12641     ชาย  29.0          ปทุมธานี
12642    หญิง  50.0           ราชบุรี
12643     ชาย  39.0              สตูล
12644    หญิง  49.0       สมุทรปราการ
12645    หญิง  44.0            ชลบุรี
12646     ชาย  18.0       สมุทรปราการ
12647     ชาย  42.0     กรุงเทพมหานคร
12648    หญิง  44.0            ชลบุรี
12649    หญิง  52.0             ระยอง
12650    หญิง  23.0             ระยอง
12651    หญิง  29.0             ระยอง
12652    หญิง  22.0               ตาก

In [237]:
this_data_updated.isnull().any() #แทนค่าใหม่จากค่าเดิมที่เคยเป็น missing

sex                  False
age                  False
province_of_onset    False
dtype: bool

###2. แทนด้วยค่าที่เหมาะสม

*   ตามใจคนทำ
*   ค่ากลาง



#### ตามใจคนทำ

In [239]:
data_covid.head()

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[5 rows x 10 columns]

In [243]:
data_covid_fillasiwant = data_covid.fillna(value={'sex':'หญิง'}) #เรามีข้อมูลว่า เพศหญิงเยอะกว่าเพศชาย ดังนนั้นจึงแทนค่าที่ missing เป็นผู้หญิง

In [244]:
data_covid_fillasiwant.isnull().any() #sex แก้ค่า missing เลยเป็น False แต่ age ยังเป็น True

No.                      False
announce_date            False
notification_date         True
sex                      False
age                       True
nationality               True
province_of_isolation     True
province_of_onset         True
district_of_onset         True
risk                      True
dtype: bool

#### ค่ากลาง

In [246]:
data_covid['age']

0        61.0
1        74.0
2        73.0
3        68.0
4        66.0
         ... 
12648    44.0
12649    52.0
12650    23.0
12651    29.0
12652    22.0
Name: age, Length: 12653, dtype: float64

In [247]:
data_covid['age'].mean() #คำนวณเฉพาะค่าที่มีอยู่ ถ้าเอาค่าmissing มาคำนวณด้วยค่าจะลดลง

38.054748339151

In [249]:
data_covid_fillwithmean = data_covid.fillna(value={'age':data_covid['age'].mean()})

In [250]:
data_covid_fillwithmean.isnull().any() #age = False

No.                      False
announce_date            False
notification_date         True
sex                       True
age                      False
nationality               True
province_of_isolation     True
province_of_onset         True
district_of_onset         True
risk                      True
dtype: bool

### การวนลูป record ในตาราง  (```.iterrows```)

In [251]:
this_data = data_covid[['sex','age','province_of_onset']]
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [252]:
for each_row in this_data.iterrows(): #.iterrows สร้าง list ของแต่ละแถว
    # print(each_row)
    # print(each_row[1])
    # print(each_row[1]['sex'])
    if (each_row[1]['age'] <= 20) and (each_row[1]['province_of_onset'] == 'ขอนแก่น'):
        print(each_row) #สามารถวนลูปเข้าไปในตารางได้

(1466, sex                      ชาย
age                       19
province_of_onset    ขอนแก่น
Name: 1466, dtype: object)
(9333, sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 9333, dtype: object)
(9334, sex                      ชาย
age                       20
province_of_onset    ขอนแก่น
Name: 9334, dtype: object)
(10610, sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 10610, dtype: object)
(11517, sex                      ชาย
age                       12
province_of_onset    ขอนแก่น
Name: 11517, dtype: object)
(11697, sex                     หญิง
age                       17
province_of_onset    ขอนแก่น
Name: 11697, dtype: object)


### การวนลูป แบบมองตารางแพนด้าส์(pandas dataframe) เป็น numpy array  (```.iloc```)

In [253]:
for each_row in range(this_data.shape[0]):
    if (this_data.iloc[each_row,1] <= 20) and (this_data.iloc[each_row,2] == 'ขอนแก่น'):  #each_row คือ ตัวเลข
        print(each_row)
        print(this_data.iloc[each_row,:])     

1466
sex                      ชาย
age                       19
province_of_onset    ขอนแก่น
Name: 1466, dtype: object
9333
sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 9333, dtype: object
9334
sex                      ชาย
age                       20
province_of_onset    ขอนแก่น
Name: 9334, dtype: object
10610
sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 10610, dtype: object
11517
sex                      ชาย
age                       12
province_of_onset    ขอนแก่น
Name: 11517, dtype: object
11697
sex                     หญิง
age                       17
province_of_onset    ขอนแก่น
Name: 11697, dtype: object


## Quiz ตัดตารางออกมาเฉพาะปี 2021 ```announce_date``` ในปี 2021 

**Hint**
*   วนลูปหา index ของปี 2021
*   ตัดตารางมาเฉพาะ ปี 2021



In [257]:
data_covid

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 10 columns]

In [279]:
TF=list() #สร้าง list เพื่อวนลูปหาปี 2021
for each_row in data_covid.iterrows(): # วนลูปหา index ของปี 2021
    if each_row[1]['announce_date'].split('/')[2] == '2021': #เลือกคอลัม announce_date โดยใช้คำสั่ง split ในการแยกเฉพาะปีแล้วต้องเท่ากับ 2021 เท่านั้น
        TF.append(True) #ถ้าวนลูปแล้วเป็น True จะเพิ่มเข้า list
    else:
        TF.append(False) #ถ้าวนลูปแล้วเป็น False จะไม่เพิ่มเข้า list

data_covid[TF].head()

No.  ...                                               risk
6884  6885  ...                                  Cluster สมุทรสาคร
6885  6886  ...                                      Cluster ระยอง
6886  6887  ...                                        สถานบันเทิง
6887  6888  ...                                  Cluster สมุทรสาคร
6888  6889  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...

[5 rows x 10 columns]

### HW กลุ่ม 2 || เลือกข้อมูลมาเฉพาะวัน announce_date เป็นวันที่ 1-20  

In [254]:
data_covid

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 10 columns]

In [256]:
data_HW2 = list() #สร้าง list เพื่อวนลูปหา index วันที่ 1-20
for each_row in data_covid.iterrows():  
    if each_row[1]['announce_date'].split('/')[1] <= '20':  #เลือกคอลัม announce_date โดยใช้คำสั่ง split ในการแยกเอาเฉพาะวันที่ โดยค่าต้องมากกว่าหรือเท่ากับ 20
        data_HW2.append(True) #ถ้าวนลูปแล้วเป็น True จะเพิ่มเข้า list
    else:
        data_HW2.append(False) #ถ้าวนลูปแล้วเป็น False จะไม่เพิ่มเข้า list
data_covid[data_HW2]

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
32        33  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
33        34  ...              บุคลากรด้านการแพทย์และสาธารณสุข
34        35  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[5135 rows x 10 columns]

# Function ตัวช่วยใน pandas

### ```.describe()``` คำนวณค่าทางสถิติของข้อมูลที่เป็นตัวเลข

In [209]:
data_covid.describe()

No.          age
count  12653.000000  8881.000000
mean    6327.000000    38.054748
std     3652.750813    15.125178
min        1.000000     0.100000
25%     3164.000000    27.000000
50%     6327.000000    36.000000
75%     9490.000000    48.000000
max    12653.000000    97.000000

### ```.mean()``` คำนวณค่าเฉลี่ยของข้อมูลโดยไม่สนใจ missing 

In [210]:
data_covid[data_covid['sex']=='ชาย']['age'].mean()

39.46076940331987

### ```.isnull()```

In [211]:
data_covid.isnull()

No.  announce_date  ...  district_of_onset   risk
0      False          False  ...               True  False
1      False          False  ...               True  False
2      False          False  ...              False  False
3      False          False  ...               True  False
4      False          False  ...               True  False
...      ...            ...  ...                ...    ...
12648  False          False  ...              False  False
12649  False          False  ...              False  False
12650  False          False  ...              False  False
12651  False          False  ...              False  False
12652  False          False  ...              False  False

[12653 rows x 10 columns]

# next
*   .describe
*   .mean
*   HW วนลูป missing .isnull
*   ต่อตารางแกน X แกน y
*   .groupby
*   save table